src/main/java/org/truffleruby/core/regexp/TruffleRegexpNodes.java

```diff
@@ -154,7 +158,40 @@ protected RubyEncoding regexpPrepareEncoding(RubyRegexp regexp, Object matchStri
                         coreExceptions().typeErrorUninitializedRegexp(this));
             }
 
-            if (regexp.encoding == matchStringEncoding) {
+            // This set of checks optimizes for the very common case where the regexp uses the default encoding and
+            // the match string is valid and uses an ASCII-compatible encoding. This extra set of checks is not present
+            // in MRI as of 3.0.2. While they permit an early return of the method, they do not conflict with any of
+            // the branches in the fallback strategy.
```

Comment:
Could you mention we don't need to care about `regexp.options.isFixed()` since that's not possible with a regexpEncoding US-ASCII?
Actually I'm not sure that's impossible:
```
> Regexp.new("a".force_encoding("us-ascii"), Regexp::FIXEDENCODING).fixed_encoding?
=> true
> Regexp.new("a".force_encoding("us-ascii"), Regexp::FIXEDENCODING) =~ "aé"
(irb):5:in `<main>': incompatible encoding regexp match (US-ASCII regexp with UTF-8 string) (Encoding::CompatibilityError)
```
Which sounds like the shortcut for `matchStringEncoding == Encodings.UTF_8` needs to care about isFixed(), and `matchStringEncoding == Encodings.BINARY` maybe too.
Probably simpler to `if (regexp.encoding == Encodings.US_ASCII && !regexp.options.isFixed()) {` here

Reply:
Good catch. I'll add a spec for this.